### PR TITLE
Add `on_predict_{batch,epoch}_{start,end}` and `Callback.on_predict_{start,end}`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Split out from #7042 and #7127

Notes:
- `self.log` is not allowed in the predict hooks

cc: @justusschock @tchaton 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by tchaton — CHANGES_REQUESTED** [review]
LGTM !
Need to add the calls on trainer callbacks too for the new ones.

File: pytorch_lightning/callbacks/lambda_function.py [hunk 1]
```diff
@@ -79,82 +79,13 @@ def __init__(
...
-            self.on_load_checkpoint = on_load_checkpoint
-        if on_after_backward is not None:
-            self.on_after_backward = on_after_backward
-        if on_before_zero_grad is not None:
-            self.on_before_zero_grad = on_before_zero_grad
+        for k, v in locals().items():
+            if k == "self":
+                continue
+            if v is not None:
+                setattr(self, k, v)
```
Comment (Borda) [inline pytorch_lightning/callbacks/lambda_function.py:91]: very nice!

File: pytorch_lightning/callbacks/progress.py [hunk 2]
```diff
@@ -219,7 +219,7 @@ def on_test_start(self, trainer, pl_module):
     def on_test_batch_end(self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx):
         self._test_batch_idx += 1
 
-    def on_predict_start(self, trainer, pl_module):
+    def on_predict_epoch_start(self, trainer, pl_module):
```
Comment (Borda) [inline pytorch_lightning/callbacks/progress.py:222]: how about back-compatibility?

Comment (carmocca) [reply pytorch_lightning/callbacks/progress.py:222]: Predict is in beta (not even that, we beta-release it with 1.3)

So we can break it

Comment (awaelchli) [reply pytorch_lightning/callbacks/progress.py:222]: I didn't even realize it was beta

File: pytorch_lightning/callbacks/lambda_function.py [hunk 3]
```diff
@@ -79,82 +79,15 @@ def __init__(
...
-            self.on_keyboard_interrupt = on_keyboard_interrupt
-        if on_save_checkpoint is not None:
-            self.on_save_checkpoint = on_save_checkpoint
-        if on_load_checkpoint is not None:
-            self.on_load_checkpoint = on_load_checkpoint
-        if on_after_backward is not None:
-            self.on_after_backward = on_after_backward
-        if on_before_zero_grad is not None:
-            self.on_before_zero_grad = on_before_zero_grad
+        for k, v in locals().items():
```
Comment (tchaton) [inline pytorch_lightning/callbacks/lambda_function.py:89]: Awesome !

File: tests/trainer/logging_/test_logger_connector.py [hunk 4]
```diff
@@ -370,6 +382,7 @@ def test_call_back_validator(tmpdir):
             assert func_name in not_supported
             with pytest.raises(MisconfigurationException, match="function doesn't support"):
                 validator.check_logging_in_callbacks(current_hook_fx_name=func_name, on_step=on_step, on_epoch=on_epoch)
+                print(func_name)
```
Comment (tchaton) [inline tests/trainer/logging_/test_logger_connector.py:385]: Mind removing this ?

**Review by tchaton — APPROVED** [review]
LGTM ! Thanks.

File: tests/trainer/logging_/test_logger_connector.py (hunk 4 above)
Comment (carmocca) [reply tests/trainer/logging_/test_logger_connector.py:385]:
```suggestion
```

File: tests/callbacks/test_lambda_function.py [hunk 5]
```diff
@@ -29,33 +29,34 @@ def on_train_epoch_start(self):
...
         limit_val_batches=1,
         callbacks=[LambdaCallback(**hooks_args)],
     )
-    results = trainer.fit(model)
-    assert results
+    trainer.fit(model)
 
-    model = CustomModel()
-    ckpt_path = trainer.checkpoint_callback.best_model_path
+    # raises KeyBoardInterrupt, also loads from checkpoint
```
Comment (carmocca) [inline tests/callbacks/test_lambda_function.py:47]:
```suggestion
    # raises KeyboardInterrupt and loads from checkpoint
```

File: pytorch_lightning/callbacks/progress.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/progress.py:222]: Added it back :)

**Pull request merged**